### DISCUSSION TITLE
tbd4vcv rack beta win64

### DISCUSSION THREAD

**Discussion by ctag-fh-kiel** [ctag-fh-kiel/ctag-tbd#45] [Announcements]
Please try the windows tbd4vcv [beta](https://github.com/ctag-fh-kiel/ctag-tbd/releases/tag/tbd4vcv-beta-win64)

copy file in "$RACK_DIR/plugins-v1" folder

once you have created an instance of tbd in rack you can change the parameters through a web server on localhost:3000 just like in the simulator. the active tbd instance linked to the web server can be selected through a context menu in vcv rack (ctrl-click on the module and activate web server).

comments in this discussion pls

**Comment by Visuelle-Musik** [comment]
Works great - I really enjoy to have my hardware TBD as a virtual module in VCV Rack too, now - thanks so much for sharing!

**Comment by ctag-fh-kiel** [comment]
Here is the latest binary from CI
[tbd4vcv-1.0.0-win64.zip](https://github.com/ctag-fh-kiel/ctag-tbd/files/7492377/tbd4vcv-1.0.0-win64.zip)

**Comment by Visuelle-Musik** [comment]
Hi, tested the newest version, but sadly it crashes VCV Rack on load, which was no issue at all before.
So the TBD4VCV can be selected form the modules-menu, but when I try to put it inside the Rack, VCV Rack is halted and a crash-dialogue-box gets displayed.
(BTW: Signal 22 stands for: **SIGTTOU "Background write to tty"**)
...
[15.465 info src/patch.cpp:86] Saving patch C:\Users\Mathias\DOCUME1/Rack/autosave-v1.vcv
[15.467 info src/settings.cpp:189] Saving settings C:\Users\Mathias\DOCUME1/Rack/settings-v1.json
[21.466 fatal src/main.cpp:45] Fatal signal 22. Stack trace:
30: ZN4rack10appDestroyEv 0x456435
29: raise 0x990f6140
28: abort 0x990f99d0
27: assert 0x990f6680
26: spi_flash_emu_init 0x4e3de240
25: ZN7tbd4vcvC1Ev 0x4ea3fc60
24: ZZN4rack11createModelI7tbd4vcv13tbd4vcvWidgetEEPNS_6plugin5ModelERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEN6TModel18createModuleWidgetEv 0x4eb7a940
23: ZN4rack3app11AudioWidget12setAudioPortEPNS_5audio4PortE 0x49839c
22: ZN4rack3app8ModelBox8onButtonERKNS_5event6ButtonE 0x75cbc0
21: ZN4rack6widget6Widget8onButtonERKNS_5event6ButtonE 0x7726f0
20: ZN4rack6widget6Widget8onButtonERKNS_5event6ButtonE 0x7726f0
19: ZN4rack6widget6Widget8onButtonERKNS_5event6ButtonE 0x7726f0
18: ZN4rack2ui12ScrollWidget8onButtonERKNS_5event6ButtonE 0x4a3ed4
17: ZN4rack6widget12OpaqueWidget8onButtonERKNS_5event6ButtonE 0x771cd0
16: ZN4rack3app14BrowserOverlay8onButtonERKNS_5event6ButtonE 0x7530d0
15: ZN4rack6widget12OpaqueWidget8onButtonERKNS_5event6ButtonE 0x771cd0
14: ZN4rack5event5State12handleButtonENS_4math3VecEiii 0x4576c8
13: ZN4rack7updater17isUpdateAvailableEv 0x45914e
12: glfwPlatformUnlockMutex 0x4d78ae
11: DispatchMessageW 0x9b1723a0
10: CallWindowProcW 0x9b173c30
9: glDebugEntry 0x77b71490
8: DispatchMessageW 0x9b1723a0
7: NotifyWinEvent 0x9b172120
6: glfwPlatformPollEvents 0x4daf66
5: ZN4rack6Window3runEv 0x4594ea
4: main 0x781840
3: main 0x781840
2: main 0x781840
1: BaseThreadInitThunk 0x998913f0
0: RtlUserThreadStart 0x9b5454c0

**Comment by Visuelle-Musik** [comment]
I built TBD4VCV from source too, which in principle worked great (after I exchanged the C-runtime for Rack), but I have the following issue:
When I end VCV Rack and start it again, there is an **error-message, that another instance of VCV Rack is already running...**
-> When I kill the task of VCV Rack that is still visiable in the windows-task-manager, I can start VCV Rack again and use (the locally built version of) TBD4VCV again. After I temporarily deleted TBD4VCV the issue went away.
So I guess that there is some tydying up / destructor-call missing/not working in TBD4VCV which prevents Rack.exe to be ended properly.
(At least this was the case, when I had the identical issue with a plugin for VCV Rack I built myself once)
But... that issue was also not present with the very first TBD4VCV which was built via Github-actions (using Docker), so maybe it's connected to the Standard GNU C / C++ libraries that got added statically?

**Reply by Visuelle-Musik** [reply]
I could fix the issue reported above myself, here is what I did:

1) I commented out a line in CMakeLists.txt, which I added there because of missing symbols in the C/C++-Runtime
-> target_link_options(${LIB_NAME} PRIVATE -static-libgcc -static-libstdc++)

2) I copied the following dll-files from C:\msys64\mingw64\bin to my VCV Rack installation folder 
20.10.2021  21:01            83.635 libgcc_s_seh-1.dll
20.10.2021  21:01         1.965.780 libstdc++-6.dll
22.10.2021  01:45            58.621 libwinpthread-1.dll

So it seems, that the objects contained in the dlls can not be unloaded on exit when linked in statically.
In result there will be an error "Rack already running" on next start
On the other hand, with the C/C++ runtime libraries distributed with VCV Rack 1.16 the compile&link of TBD4VCV will not work properly when msys64/mingw64 is used as suggested on the VCV Rack site.
So this is a kind of catch 22 that hopefully will become obsolete with VCV Rack 2.x?

But the important thing is, that the issue is not present with the version built under Ubuntu/Docker for Windows on Github!

**Comment by Visuelle-Musik** [comment]
There is another issue with the locally built version of TBD4VCV, which I noticed just now:
**New TBD-plugin will not show up**, when I exchange "**spm-config.jsn**" with the initial version of this file under 
**C:\Users\\'YourName'\Documents\Rack\plugins-v1\tbd4vcv\spiffs_image\data**, the new plugin still will not show up at localhost:3000/
But this works with the identical source-code with the simulator and exchanging "spm-config.jsn" with the path 
C:\ctag-tbd-vcv\ctag-tbd\tbd4vcv\build\tbd4vcv\spiffs_image\data and the browser pointing to localhost:8080/

**Reply by ctag-fh-kiel** [reply]
spm-config.jsn and any of its settings aren't currently used in the plugin.

**Reply by Visuelle-Musik** [reply]
Ok, that's what I thought by now, too.
So, what would be the correct way to add a TBD-Plugin to TBD4VCV then?
I also updated spm-config.jsn in the main setting of the vcv branch, and can run the simulator with this and my (new) plugin shows up in the simulator, but not in VCV Rack?

**Reply by ctag-fh-kiel** [reply]
Recompiling and packaging + dropping the .zip file in plugins-v1 + erasing the old folder.
Plus starting a fresh patch in VCV Rack. 
The plugin data is stored in the "patch.vcv" file, just open it, it's a json and includes the plugin parts, that usually are in spm-config.jsn.
This way it is possible to manage multiple plugin instances.

**Reply by Visuelle-Musik** [reply]
Great - now it works.
The only thing I had not done (yet) was to start a fresh patch.
So I guess there are some caching-mechanisms in VCV Rack that have to be overwritten that way?

**Reply by ctag-fh-kiel** [reply]
This behaviour could be changed, so that every time the plugin is initiated all currently available plugins are scanned.
Does this make better sense?

**Reply by Visuelle-Musik** [reply]
mh - for most users this would be many plugins to be scanned. I guess it depends on the time it takes.
I could test the behavour, my PC is quite old/slow, plugins being on an SDD, though...
Would the same thing (with the same amount of effort) be possible with a menu-item  "rescan TBD plugins" with TBD4VCV, maybe?

**Comment by ctag-fh-kiel** [comment]
I have updated the binary in the release, it should work now.

**Comment by Visuelle-Musik** [comment]
Great - I can confirm, that the crash of VCV Rack is gone with the new binary of TBD4VCV!

I tried with the original C/C++ Runtime Windows-dlls in the Rack-directory, as well as the (newer) GNU versions and there is no difference noticable!

**Reply by ctag-fh-kiel** [reply]
Awesome :)
